File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        parent = dict()
        depth = dict()
        def res(node = root, par = None, d = 0):
            if node is not None:
                parent[node] = par
                depth[node] = d
                res(node.left, node, d + 1)
                res(node.right, node, d + 1)
        res()
        
        while p != q:
            if depth[p] > depth[q]:
                p = parent[p]
            elif depth[p] < depth[q]:
                q = parent[q]
            else:
                p = parent[p]
                q = parent[q]
        
        return p
        

File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode s) {
        Map<TreeNode, TreeNode> parent = new HashMap<>();
        Map<TreeNode, Integer> level = new HashMap<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        level.put(root, 0);
        while (q.size() != 0) {
            TreeNode u = q.poll();
            if (u.left != null) {
                q.add(u.left);
                parent.put(u.left, u);
                level.put(u.left, level.get(u) + 1);
            }
            if (u.right != null) {
                q.add(u.right);
                parent.put(u.right, u);
                level.put(u.right, level.get(u) + 1);
            }
        }
        while (level.get(p) > level.get(s)) {
            p = parent.get(p);
        }
        while (level.get(s) > level.get(p)) {
            s = parent.get(s);
        }
        while (p != s) {
            p = parent.get(p);
            s = parent.get(s);
        }
        return p;
    }
}